### PR TITLE
Adding const-lr and const-lr-cooldown LR schedulers

### PR DESCRIPTION
**Pull request by JeniaJitsev**
Adding const-lr and const-lr-cooldown LR schedulers for training with const lr followed by a cooldown schedule. 

Motivation is to have a training duration independent schedule. With this, having only a single full training run for max samples seen can be for instance used for performance evaluation at any samples seen smaller than the max, without repeating the experiment (contrary to training duration dependent schedules, eg cosine annealing). 

Also, training can be continued to arbitrary sample seen scale from any checkpoint when using training duration independent schedules.

Additional options are:

--lr-scheduler "cosine": cosine annealing (current default)

--lr-scheduler "const" : using const learning rate (together with warm up schedule if defined)

--lr-scheduler "const-cooldown" : using const learning rate followed by a cooldown schedule (linear and polynomial schedules currently supported)
Further options for const-cooldown:
--lr-cooldown-epochs M : cooldown for fixed number of M epochs after running const lr; eg if specifying total N epochs, there will be N-M epochs executed with const lr, followed by M epochs with cooldown schedule
--lr-cooldown-end val : cooling down to learning rate = val; default is 0
--lr-cooldown-power val : linear decay cooldown schedule for 1 (default); polynomial for > 1 values.

### CODE REVIEW

File: src/training/params.py [hunk 1]
```diff
@@ -124,6 +124,15 @@ def parse_args(args):
...
     )
+    parser.add_argument(
+        "--lr-scheduler",
+        type=str,
+        default='cosine',
+        help="LR scheduler. Options: cosine (cosine schedule, default), const (constant lr), const-cooldown (constant lr followed by a cooldown)",
+    )
+    parser.add_argument("--end-lr", type=float, default=0.0, help="End learning rate for cooldown schedule. Default: 0")
+    parser.add_argument("--power-lr", type=float, default=1.0, help="Power for polynomial cooldown schedule. Default: 1.0 (implementing simple linear decay)")
+    parser.add_argument("--cooldown-epochs", type=int, default=None, help="Epochs to perform cooldown for cooldown schedule. Will start performing cooldown from total_epochs - cooldown_epochs on.")
```
[the remark this answers is not in the document]
Comment (rwightman) [reply src/training/params.py:135]: addition to that, if `--lr-end` applies to all schedules, (similarly called `min`) then we can leave the cooldown out, but if it only applies to cooldown and cosine does not use it to change the end LR it should have cooldown in the name... it's also not uncommon to have a constant cooldown after a cosine or other sched FYI

**Review by rwightman — CHANGES_REQUESTED** [review]

File: src/training/main.py [hunk 2]
```diff
@@ -273,7 +275,16 @@ def main(args):
...
-        scheduler = cosine_lr(optimizer, args.lr, args.warmup, total_steps)
+        if args.lr_scheduler == "cosine-annealing":
+            scheduler = cosine_lr(optimizer, args.lr, args.warmup, total_steps)
+        elif args.lr_scheduler == "const-lr":
+            scheduler = const_lr(optimizer, args.lr, args.warmup, total_steps)
+        elif args.lr_scheduler == "const-lr-cooldown":
+            assert args.cooldown_epochs is not None, "Please specify the number of cooldown epochs for this lr schedule."
+            cooldown_steps = (data["train"].dataloader.num_batches // args.accum_freq) * args.cooldown_epochs
+            scheduler = const_lr_cooldown(optimizer, args.lr, args.warmup, cooldown_steps, args.power_lr, args.end_lr, total_steps)
+        else:
```
[the remark this answers is not in the document]
Comment (JeniaJitsev) [reply src/training/main.py:286]: Previously, the method was called "cosine_lr" (used by default). You suggest we introduce new naming convention for the methods then?

File: src/training/main.py [hunk 3]
```diff
@@ -273,7 +275,16 @@ def main(args):
...
+        if args.lr_scheduler == "cosine":
+            scheduler = cosine_lr(optimizer, args.lr, args.warmup, total_steps)
+        elif args.lr_scheduler == "const":
+            scheduler = const_lr(optimizer, args.lr, args.warmup, total_steps)
+        elif args.lr_scheduler == "const-cooldown":
+            assert args.cooldown_epochs is not None, "Please specify the number of cooldown epochs for this lr schedule."
+            cooldown_steps = (data["train"].dataloader.num_batches // args.accum_freq) * args.cooldown_epochs
+            scheduler = const_lr_cooldown(optimizer, args.lr, args.warmup, cooldown_steps, args.power_lr, args.end_lr, total_steps)
+        else:
+            print(f"Unknown scheduler, {args.lr_scheduler}")
```
[the remark this answers is not in the document]
Comment (JeniaJitsev) [reply src/training/main.py:287]: I would suggest to exit with an error - as default as cosine, people should get termination as most probably they were providing a name that is not in the schedule list and it will probably be confusing if that leads to execution of cosine schedule - which was not provided in this scenario as  a choice.

**Review by JeniaJitsev — COMMENTED** [review]
Made the changes in the PR

**Review by rwightman — APPROVED** [review]

**Comment by rwightman** [comment]
merged up to the last log which was equivalently addded
